src/BuiltInTools/dotnet-watch/Utilities/ProjectGraphNodeExtensions.cs

```diff
@@ -31,4 +31,28 @@ public static bool IsNetCoreApp(this ProjectGraphNode projectNode, Version minVe
 
     public static IEnumerable<string> GetCapabilities(this ProjectGraphNode projectNode)
         => projectNode.ProjectInstance.GetItems("ProjectCapability").Select(item => item.EvaluatedInclude);
+
+    public static IEnumerable<ProjectGraphNode> GetTransitivelyReferencingProjects(this IEnumerable<ProjectGraphNode> projects)
+    {
+        var visited = new HashSet<ProjectGraphNode>();
+        var queue = new Queue<ProjectGraphNode>();
+        foreach (var project in projects)
+        {
+            queue.Enqueue(project);
+        }
+
+        while (queue.Count > 0)
+        {
+            var project = queue.Dequeue();
+            if (visited.Add(project))
```

Comment:
I'm not familiar with ProjectGraphNode, but is it guaranteed that either:
* A single ProjectGraphNode object will always represent the project from different references?
* or ProjectGraphNode overrides GetHashCode() and Equals() to allow similar ProjectGraphNodes to match?

Reply:
It would be odd if new instances of ProjectGraphNode were created as we traverse the graph.